<a href="https://colab.research.google.com/github/jjcrofts77/TMB-MATH34041/blob/main/content/notebooks/Chapter0/PythonBasics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 0.1 Basics in Python

This section aims to cover simple, core programming concepts in Python.

## Installing Python and add-ons
If you wish to install Python on your own computer it is advisable to use an integrated system (IS) to avoid compatability issues. For this course the recommended IS is [Anaconda](https://www.anaconda.com/), which is free to use and ships with [Spyder](https://www.spyder-ide.org/) a scientific Python development environment (see {numref}`FigSpyderIDE3` for an example of the Spyder environment).

```{figure} ../../images/SpyderIDEExample.png
---
height: 450px
name: FigSpyderIDE3
---
Illustration of a Python script as seen in the Anaconda Spyder IDE
```

Once installed, you can launch Spyder through the Anaconda Navigator, which can be found under the list of programs prompted by opening Windows Start menu.

The IDE consists of an Editor Window, an Object Inspector Window and The Console. The Editor Window is where we write our Python scripts; the Object Inspector is where we browse files, explore variables in the workspace and visualise plots; whilst the Console is where we access the iPython shell and see the results of our scripts.

To create a Python script we enter commands into the editor, save our progam and run it by clicking on the green triangle in the upper-lefthand corner of the IDE; the results will displayed in the Python console in the lower-righthand pane of the IDE. (Again, see {numref}`FigSpyderIDE3`)

## Google Colab

Google Colab, or Colaboratory, is a free Jupyter notebook environment that runs in the cloud. It allows you to write and execute Python code, as well as other programming languages, in a web browser. Colab notebooks are a great way to combine code, text, and images in a single document. This makes it easy to document your work and share it with others.

Colab has two types of cells: text and code. Text cells are formatted using a simple markup language called Markdown.

See the following link for a brief description of the [Google Colab Markdown](https://colab.research.google.com/notebooks/markdown_guide.ipynb#scrollTo=Lhfnlq1Surtk) langauge.

## Python as a calculator
Python can be used as a calculator to perform basic and advanced arithmetic operations. Here are some examples of how to use Python as a calculator:

  **Basic arithmetic operations:** Python can be used to perform basic arithmetic operations such as addition, subtraction, multiplication, and division. For example, the following code would add two numbers and print the result:

In [ ]:
a = 10
b = 5

print(a + b)

15


  **Advanced arithmetic operations:** Python can also be used to perform more advanced arithmetic operations such as exponentiation, logarithms, and trigonometric functions. For example, the following code would calculate the square root of a number and print the result:

In [ ]:
import math

number = 16

print(math.sqrt(number))

4.0


In the above we have imported the Python math module which contains the standard mathematical functions (*e.g.* $\cos, \sin, \exp, \sinh, \cosh$ *etc.*).


## Basic programming in Python

In this section we shall see how Python can be used to write simple Python programs. We shall focus on three programming structures:

  1. defining functions
  2. using for and while loops
  3. if, elif, else constructs

We have seen versions of these structures when learning Matlab in Years 1 and 2.

### Defining functions
Below I have written a function for converting degrees Celsius to Kelvin.


In [ ]:
# A function to convert degrees Celsius to Kelvin

def C2K():
  C = int(input('Enter temperature in degrees Celsius: '))
  K = C + 273.15
  print('Temperature in Kelvin is {} K'.format(K))

# run the function
C2K()

Enter temperature in degrees Celsius: 25
Temperature in Kelvin is 298.15 K


Note that the function above did not require any inputs.

Suppose we want to write a program that defines the *logistic equation* given by

$$f(x) = \mu x(1-x)$$

Then we would define a function as follows.

In [ ]:
# The logistic equation - save as f_mu.py

def f_mu(mu, x):
  return mu*x*(1-x)

# run the function for a given input
mu, x = 1, 0.25
fm = f_mu(mu, x)

# print result
print('fm = {}'.format(fm))

fm = 0.1875


### Using loops
The following examples illustrate how for and while loops can be used to speed up repetitive tasks.

Let us write a program that lists the first $n$ terms of the Fibonacci sequence

$$
x_{k+2} = x_{k+1}+x_k: \quad x_0 = 0,~~ x_1 = 1
$$


In [ ]:
# A function to list the first n terms of the Fibonacci sequence

def fibonacci(n):
  a, b = 0, 1
  print(a)
  print(b)
  print(a+b)
  for i in range(n-3):
    a, b = b, a+b
    print(a+b)

# run function
fibonacci(20)

0
1
1
2
3
5
8
13
21
34
55
89
144
233
377
610
987
1597
2584
4181


Recall that the above Fibonacci sequence can be solved analytically to obtain

$$
x_k = \frac{\phi^k-\psi^k}{\sqrt{5}}
$$

where

$$
\phi = \frac{1+\sqrt{5}}{2}
$$

is the *golden ratio* and $\psi=1-\phi$ is its conjugate.

We can check our result using this formula. For example, for $n=20$ we have that

$$
x_{20} = \frac{\phi^{20}-\psi^{20}}{\sqrt{5}} = 4181
$$

which agrees with our Python code.

Next lets use loops to sum the first $n$ natual numbers

In [ ]:
# A function to sum the first n natural numbers

def sum_n(n):
  sum = 0
  i = 1
  while i <= n:
    sum += i
    i += 1   # increment counter
  print('The sum is', sum)

# run function
sum_n(100)

The sum is 5050


Again this is easy to check since we know that

$$
\sum_{i=1}^{100}i = \frac{100(101)}{2} = 5050
$$

as expected.

### Conditional statements using if, elif, else

Lets write a program that returns a students grade for the Maths Bio coursework given a score out of 100.

In [ ]:
# A program to grade the Topics in Mathematical Biology coursework

def grade(score):
  if score >= 92.5:
    GBA = '1EXC'
  elif score >= 84.5:
    GBA = '1HIGH'
  elif score >= 77.5:
    GBA = '1MID'
  elif score >= 69.5:
    GBA = '1LOW'
  else:
    GBA = 'Fail'
  return GBA

# run the function
grade(94)

'1EXC'

What about a game of guess the number!

In [ ]:
# Guess the number

import random # Import the random number module

num_guesses = 0
name = input('Hi! What is your name? ')
number = random.randint(1, 20) # a random number between 1 and 20
print('Welcome, {}! I am thinking of an integer between 1 and 20.'.format(name))

while num_guesses < 6:
  guess = int(input('Take a guess and type the integer '))
  num_guesses += 1

  if guess < number:
    print('Your guess is too low')
  if guess > number:
    print('Your guess is too high')
  if guess == number:
    break

if guess == number:
  print('Damn {}, you win! You guessed the correct number in {} guesses'.format(name, num_guesses))
else:
  print('You lose! The number I was thinking of was {}'.format(number))

Hi! What is your name? Jonathan
Welcome, Jonathan! I am thinking of an integer between 1 and 20.
Take a guess and type the integer 1
Your guess is too low
Take a guess and type the integer 2
Your guess is too low
Take a guess and type the integer 12
Your guess is too high
Take a guess and type the integer 6
Your guess is too high
Take a guess and type the integer 9
Your guess is too high
Take a guess and type the integer 7
Your guess is too high
You lose! The number I was thinking of was 3


In the above program we have imported a module in order to access functions for simulating random numbers. In order to use Python to study networks we are going to have to import other modules and libraries as will be demonstrated in the sections to follow.

### Useful data types

#### **Lists**
A Python *list* is an *ordered*, *mutable* collection of non-unique items. Since a list is ordered we can call entries of the list by their index.

In [72]:
# list example

student_names = ['Ruth','Theo','Benjamin','Vindula']
print('The second student in the list is ' + student_names[1] + ' and the last is ' + student_names[-1] + '\n')

# slice a list
print('Slice a list as follows:\n')
print(student_names[2:4])

The second student in the list is Theo and the last is Vindula

Slice a list as follows:

['Benjamin', 'Vindula']


By *mutable* we mean that the list can be changed by adding or removing items. For example, to add items to the end of the list we can use the .append() command. Note that items on the list do not need to be unique.

In [73]:
# I forgot to add Sam to the list :)
student_names.append('Sam')
print('Students enrolled on MATH34041/44041 in 2023/24 academic year:\n')
print(student_names)


# say a warm welcome to all students on the course (loop through list)
print('\n A warm welcome to all students on the course this year! \n')
for student_name in student_names:
  print('Hello ' + student_name + '!')

Students enrolled on MATH34041/44041 in 2023/24 academic year:

['Ruth', 'Theo', 'Benjamin', 'Vindula', 'Sam']

 A warm welcome to all students on the course this year! 

Hello Ruth!
Hello Theo!
Hello Benjamin!
Hello Vindula!
Hello Sam!


#### **Tuples**
*Tuples* are similar to lists. They are also ordered collections of non-unique items. The key difference being that they are *immutable*.

In [74]:
# tuple example

student_grade = ('Theo', 'Topics in Mathematical Biology', 'FMarg')
student_grade

('Theo', 'Topics in Mathematical Biology', 'FMarg')

Since they are immutable, calling the student_grade.append() command would return an error.

In [75]:
student_grade.append('N0925938')

AttributeError: 'tuple' object has no attribute 'append'

This property of tuples means that they are useful for *unpacking* since we can be sure that index properties remain fixed (e.g. student_grade[0] will always return the student name).

In [76]:
student_name, subject, grade = student_grade

print('Name: ' + student_name)
print('Course: ' + subject)
print('Grade: ' + grade)

Name: Theo
Course: Topics in Mathematical Biology
Grade: FMarg


This approach can be useful when using loops.

In [77]:
# THIS IS A TRUE TUPLE. The events depicted in this tuple took place in Nottingham in 2024. At the request of the students,
# the marks have been changed. Out of respect for the dead, the rest has been told exactly as it occurred.
student_grades = [
    ('Ruth', 'Topics in Mathematical Biology', '1LOW'),
    ('Theo', 'Topics in Mathematical Biology', 'FMarg'),
    ('Benjamin', 'Topics in Mathematical Biology', '1MID'),
    ('Vindula', 'Topics in Mathematical Biology', '21HIGH'),
    ('Sam', 'Topics in Mathematical Biology', '22MID'),
]

for student_name, subject, grade in student_grades:
  if grade.startswith('1'):
    print('Congratulations', student_name,
          'on getting a',grade,
          'in', subject,'\n')


Congratulations Ruth on getting a 1LOW in Topics in Mathematical Biology 

Congratulations Benjamin on getting a 1MID in Topics in Mathematical Biology 




#### **Dictionaries**
The last type of collection we consider is the *dictionary* which is an *unordered*, *mutable* collection of *unique* items. Below we give a very simple example.

In [78]:
# dictionary example
foreign_langauges = {
    'Ruth': 'Spanish',
    'Theo': 'English',
    'Vindula': 'Italian',
    'Sam': 'French',
}

foreign_langauges['Sam']

'French'

We can check whether a particular *key* (names in our example) is in a dictionary.

In [79]:
'Martin' in foreign_langauges

False

In [80]:
'Theo' in foreign_langauges

True

We can add, delete and change entries in the dictionary.

In [81]:
foreign_langauges['Martin'] = 'Klingon'
print('Lets add Martin and his preferred langauge: ', foreign_langauges,'\n')

# add Benjamin since he is missing from the list
foreign_langauges['Benjamin'] = 'French'

# and lets remove Martin :)
del foreign_langauges['Martin']
print('Updated and corrected list:', foreign_langauges)

Lets add Martin and his preferred langauge:  {'Ruth': 'Spanish', 'Theo': 'English', 'Vindula': 'Italian', 'Sam': 'French', 'Martin': 'Klingon'} 

Updated and corrected list: {'Ruth': 'Spanish', 'Theo': 'English', 'Vindula': 'Italian', 'Sam': 'French', 'Benjamin': 'French'}
